WIP: support for multiple KDCs (untested)

lhoss opened the pull request:
Fixes https://github.com/ultratendency/kerberos_server/issues/21 .

Status: WIP 
Cases:
* default: should work (lets see the outcome of travis tests)
* multiple KDCs: not fully tested yet

NadOby commented:
Hi @lhoss Can you please take a look on the PR, and make it pass CI please.
The problem is lint failure in molecule check

lhoss commented:
> The problem is lint failure in molecule check

I just fixed the lint issues in my changes .. but now there are 4 new lint issues on existing code (maybe due to some newer pip module ?!)
Those should be solved in a different PR  cc @HorizonNet .

Anyway I'll now continue work on this PR, mainly testing if it works with a replica kdc 👍

NadOby commented:
@lhoss  I opened PR to your rep with all fixes implemented so you can  just pick 
https://github.com/scigility/kerberos_server/pull/1

lhoss commented:
@NadOby great work thx (and so quick)!
To get some more order, I propose following:
* you open a PR to the upstream `ultratendency` repo, based on `master` (Adv.: it has travis enabled!)
  * or we can work on our `scigility` repo, then I'ld rebase your PR to master.. and then I merge it..
* and I'll rebase my (WIP!) branch on your work
wdyt ?

NadOby commented:
I opened PR to your repo.
The proposition is to do all work in your _scigility_ repo, cause major of work was done inside it, and then update _ultratendency_ one.

lhoss commented:
>  The proposition is to do all work in your _scigility_ repo, cause major of work was done inside it, and then update _ultratendency_ one.

@NadOby thanks again for your contribution
I rebased your commits to (upstream) master, to get rid of my initial WIP(PR) commit, and opened a new PR to upstream: https://github.com/ultratendency/kerberos_server/pull/23
I will rebase
* my master repo to those commits, and suggest you do the same
* and then will rebase also the now *finished* PR https://github.com/ultratendency/kerberos_server/pull/22 (check it out, happy for a review!)

lhoss commented:
This PR is now ready for review 👍 
* I'ld suggest however to first review and merge the smaller fix PR https://github.com/ultratendency/kerberos_server/pull/23
   * and then we can re-run the molecule test scenario including the `lint` (which is is fixed in -^ PR)

lhoss commented:
This PR is now ready for review 👍 
* Note:  This PR should be merged only *after* merging PR https://github.com/ultratendency/kerberos_server/pull/23, because it contains it's commits  (since I merged that PR°23 to my fork's master already)

NadOby commented:
Hi, @lhoss I'm gone over the PR and there are errors when running molecule test on the default scenario name.
`molecule test`
You don't see it because it is commented out in .travis.yaml and rely on it to effectively check code validity.
So you can uncomment it in .travis.yaml
Excepting this small issue, all seems good.

HorizonNet commented on tasks/kprop_replication.yml:
```diff
@@ -0,0 +1,77 @@
+    _keytab: "/etc/krb5.keytab"
+  loop: "{{ kerberos_server_kdcs }}"
+  when: is_master_host
+
```
Use a single empty line.

HorizonNet commented on tasks/kprop_replication.yml:
```diff
@@ -0,0 +1,77 @@
+    dest: /etc/
+    mode: 0644
+  when: is_replica_host
+
```
Ditto.

HorizonNet commented on tasks/kprop_replication.yml:
```diff
@@ -0,0 +1,77 @@
+        state: started
+        enabled: True
+  when: is_replica_host
+
```
Ditto.

HorizonNet commented:
We should also create a new version after merging this. In preparation I tagged the current master as v1.0.0. After this one I would create a v2.0.0.

HorizonNet commented:
@lhoss Do we also need to raise the minimum Ansible version?

lhoss commented on molecule/kdc-with-replica/molecule.yml:
```diff
@@ -0,0 +1,73 @@
+# Using geerlingguy's image that provides centos7 with ansible AND systemd
+platforms:
+  - name: "${MOLECULE_DISTRO:-centos7}-01.local"
+    image: "geerlingguy/docker-${MOLECULE_DISTRO:-centos7}-ansible:latest"
```
I'ld prefer the new one from geerlingguy because we can avoid to (consume time to) build the custom image. Now pushing a commit to change the default scenario

lhoss commented on .travis.yml:
```diff
@@ -9,10 +9,14 @@ services:
   - pip install docker
 script:
-  - molecule test
+  # TODO re-enable the simple test after we merged the molecule lint fix PR
```
Pushed that change, and of course it broke the build
https://travis-ci.com/scigility/kerberos_server/jobs/216113647#L908

(even though it creates more work)The fix should be done in the parent PR #23 , see my comment there https://github.com/ultratendency/kerberos_server/pull/23/files#r303446450, and then this PR rebased

lhoss commented:
> @lhoss Do we also need to raise the minimum Ansible version?

Yes  ansible 2 is a more than a must, but I recommend to put >2.5 (I ran a test with 2.5.8 and that worked 👍 )

lhoss commented:
@HorizonNet I fixed all your (good!) review comments.
Now (after lunch) I want to (force)push the latest state, rebased on #23 improvements, which will also make both molecule test scenarios (the simple/default one, and the `kdc-with-replica`) work. I hope you do not mind (github might not show the comment history after the force push)
After that I also want to push a small improvement in the  `add_princ_and_kt.yml`, making it flexibler to re-use that outside the role (plus a keytab-based kinit 1-task include, which I often need in playbooks, and which is not worth to put into a separate role, but I like to abstract it out anyway 👍 )

lhoss commented:
ps: @HorizonNet wdyt about this new kinit checks feature, that I created in a derived branch?!
https://github.com/scigility/kerberos_server/compare/support_multiple_kdcs...scigility:support_multiple_kdcs-with_kinit

For me it is useful to re-use in other playbooks, and to avoid to duplicate the logic (even though it is only 1 task , I know). Here an example with import_role:
```
  tasks:
    - import_role:
        name: ../../kerberos_server
        #name: kerberos_server
        tasks_from: kinit_kt
      vars:
        # ensure kerberos vars are set, that align with kerberos_server vars !
        _realm_name: "{{ kerberos_server_realm_name }}"
        _principal: "{{ hdfs_root_user }}/{{ ansible_fqdn }}"
        _keytab: "{{ hdfs_keytab }}"
        _kinit_user: "{{ hdfs_root_user }}"
      when: do_kinit | default(True) | bool
      tags: kinit
```
